File: binding-python/runtime/src/main/python/etch/binding/transport/fmt/binary/BinaryTaggedData.py
"""
#  Licensed to the Apache Software Foundation (ASF) under one   *
#  or more contributor license agreements.  See the NOTICE file *
#  distributed with this work for additional information        *
#  regarding copyright ownership.  The ASF licenses this file   *
#  to you under the Apache License, Version 2.0 (the            *
#  "License"); you may not use this file except in compliance   *
#  with the License.  You may obtain a copy of the License at   *
#                                                               *
#    http://www.apache.org/licenses/LICENSE-2.0                 *
#                                                               *
#  Unless required by applicable law or agreed to in writing,   *
#  software distributed under the License is distributed on an  *
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
#  KIND, either express or implied.  See the License for the    *
#  specific language governing permissions and limitations      *
#  under the License. 
"""
from __future__ import absolute_import

import types
from .....python.Exceptions import *
from .....python.Types import *
from ..TypeCode import TypeCode
from ..TaggedData import *
from ...ArrayValue import *
from ....msg.StructValue import *

# TODO - make this singleton
class NoneObject(object):
    def __repr__(self):
        return 'NONE'

class BinaryTaggedData(TaggedData):
    """
    Common elements for binary tagged data input and output
    """
    NONE = NoneObject()
    """Sentinal which marks the end of a struct or array item list in a binary i/o stream."""

    VERSION = 1
    """This is the current version of the protocol"""

    def __init__(self, valueFactory):
        """
        Constructs the BinaryTaggedData
        """
        super(BinaryTaggedData,self).__init__(valueFactory)

    def adjSmallValue(self, typeCode, value):
        if typeCode in (TypeCode.BYTE1, TypeCode.SHORT1, TypeCode.INT1, TypeCode.LONG1):
            # value = value.byteValue
            v = Byte(value)
            if v >= TypeCode.MIN_SMALL_INT and v <= TypeCode.MAX_SMALL_INT:
                if v >= 0:
                    if typeCode == TypeCode.BYTE1:
                        return Byte(v + TypeCode.POSITIVE_BYTE_BASE)
                    elif typeCode == TypeCode.SHORT1:
                        return Byte(v + TypeCode.POSITIVE_SHORT_BASE)
                    elif typeCode == TypeCode.INT1:
                        return Byte(v + TypeCode.POSITIVE_INT_BASE)
                    else:
                        return Byte(v + TypeCode.POSITIVE_LONG_BASE)

                if typeCode == TypeCode.BYTE1:
                    return Byte(-(v+1) + TypeCode.NEGATIVE_BYTE_BASE)
                elif typeCode == TypeCode.SHORT1:
                    return Byte(-(v+1) + TypeCode.NEGATIVE_SHORT_BASE)
                elif typeCode == TypeCode.INT1:
                    return Byte(-(v+1) + TypeCode.NEGATIVE_INT_BASE)
                else:
                    return Byte(-(v+1) + TypeCode.NEGATIVE_LONG_BASE)

        return typeCode

    def checkValue(self, value, validator=None):

        if not isinstance(value,(int,long)) and not value:
            return TypeCode.NULL

        if value == BinaryTaggedData.NONE:
            return TypeCode.NONE

        if validator:
            typeCode = validator.checkValue(value)
            if not typeCode:
                raise UnsupportedOperationException, "typeCode is NULL"

            return self.adjSmallValue(typeCode, value)

        if isinstance(value, (types.IntType, types.LongType, types.FloatType)):
            if isinstance(value, Byte):
                return BinaryTaggedData.checkByte(value, TypeCode.BYTE1, TypeCode.BYTE_DT)
            if isinstance(value, Short):
                return BinaryTaggedData.checkShort(value, TypeCode.SHORT2, TypeCode.SHORT_DT)
            if isinstance(value, types.IntType):
                return BinaryTaggedData.checkInteger(value, TypeCode.INT4, TypeCode.INT_DT)
            if isinstance(value, types.LongType):
                return BinaryTaggedData.checkLong(value, TypeCode.LONG8, TypeCode.LONG_DT)
            if isinstance(value, types.FloatType):
                return TypeCode.FLOAT8
            return TypeCode.CUSTOM

        if isinstance(value, types.StringTypes):
            if len(value) == 0:
                return TypeCode.EMPTY_STRING
            return TypeCode.STRING

        if isinstance(value, types.BooleanType):
            if value:
                return TypeCode.BOOLEAN_TRUE
            else:
                return TypeCode.BOOLEAN_FALSE

        if isinstance(value, StructValue):
            return TypeCode.STRUCT

        if isinstance(value, ArrayValue):
            return TypeCode.ARRAY

        if isinstance(value, ByteArray):
            return TypeCode.BYTES

        if isinstance(value, (types.ListType, types.TupleType)):
            return TypeCode.ARRAY

        return TypeCode.CUSTOM

    def getNativeTypeCode(self, clss):
        classMap = {
            types.BooleanType   : TypeCode.BOOLEAN_TRUE,
            Byte                : TypeCode.BYTE1,
            Short               : TypeCode.SHORT2,
            Integer             : TypeCode.INT4,
            Long                : TypeCode.LONG8,
            Float               : TypeCode.FLOAT4,
            Double              : TypeCode.Double,
            types.IntType       : TypeCode.INT4,
            types.LongType      : TypeCode.LONG8,
            types.FloatType     : TypeCode.FLOAT8,
            types.StringType    : TypeCode.STRING,
            types.UnicodeType   : TypeCode.STRING,
            object.__class__    : TypeCode.ANY
        }

        return classMap.has(clss, TypeCode.CUSTOM)

    def getCustomStructType(self, clss):
        return self.valueFactory.getCustomStructType(clss)

    def getNativeType(self, typeCode):
        typeMap = {
            TypeCode.BOOLEAN_TRUE   : types.BooleanType,
            TypeCode.BOOLEAN_FALSE  : types.BooleanType,
            TypeCode.BYTE1          : Byte,
            TypeCode.SHORT1         : Byte,
            TypeCode.SHORT2         : Short,
            TypeCode.INT1           : Byte,
            TypeCode.INT2           : Short,
            TypeCode.INT4           : types.IntType,
            TypeCode.LONG1          : Byte,
            TypeCode.LONG2          : Short,
            TypeCode.LONG4          : types.IntType,
            TypeCode.LONG8          : types.LongType,
            TypeCode.FLOAT4         : Float,
            TypeCode.FLOAT8         : types.FloatType,
            TypeCode.EMPTY_STRING   : types.UnicodeString,
            TypeCode.STRING         : types.UnicodeString,
            TypeCode.ANY            : object
        }

        try:
            return typeMap[typeCode]
        except:
            raise IllegalArgumentException, "Unsupported native type %s" % typeCode


    @staticmethod
    def checkByte(v, is1, dt):
        v = Byte(v)
        if v >= TypeCode.MIN_SMALL_INT and v <= TypeCode.MAX_SMALL_INT:
            if v >= 0:
                return Byte(TypeCode.PSVALUE | (Byte(dt) << TypeCode.PDTSHIFT) | v)
            return Byte(TypeCode.NSVALUE | (Byte(dt) << TypeCode.NDTSHIFT) | -(v+1))
        return is1

    @staticmethod
    def checkShort(v, is2, dt):
        if v >= Byte.MIN_VALUE and v <= Byte.MAX_VALUE:
            return BinaryTaggedData.checkByte(v, is2+1, dt)
        return is2

    @staticmethod
    def checkInteger(v, is4, dt):
        if v >= Short.MIN_VALUE and v <= Short.MAX_VALUE:
            return BinaryTaggedData.checkShort(v, is4+1, dt)
        return is4

    @staticmethod
    def checkLong(v, is8, dt):
        if v >= Integer.MIN_VALUE and v <= Integer.MAX_VALUE:
            return BinaryTaggedData.checkInteger(v, is8+1, dt)
        return is8

File: binding-go/runtime/src/main/go/etch/BinaryTaggedData.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package etch

import "fmt"
import "reflect"

const (
	VERSION int8 = 3
)

const (
	NULL          int8 = -128
	NONE          int8 = -127
	BOOLEAN_FALSE int8 = -126
	BOOLEAN_TRUE  int8 = -125
	BYTE          int8 = -124
	SHORT         int8 = -123
	INT           int8 = -122
	LONG          int8 = -121
	FLOAT         int8 = -120
	DOUBLE        int8 = -119
	BYTES         int8 = -117
	ARRAY         int8 = -111
	EMPTY_STRING  int8 = -110
	STRING        int8 = -109
	STRUCT        int8 = -108
	CUSTOM        int8 = -107
	ANY           int8 = -106
	MIN_TINY_INT  int8 = -64
	MAX_TINY_INT  int8 = 127
)

type ArrayValue struct {
	TypeCode int8
	CustomTy *Type
	Dim      int
	ArrayVal reflect.Value
}

func NewArrayValueWith(array interface{}, ty int8, cty *Type) *ArrayValue{
	ret := &ArrayValue{ty, cty,1,reflect.ValueOf(array)}
	return ret
}


func NewArrayValue(array interface{}, vf ValueFactory) *ArrayValue{
	//only support 1-dimensional, non-empty arrays up to now...
	//other arrays are likely to crash
	ret := &ArrayValue{0,nil,1,reflect.ValueOf(array)}
	Log("NewArrayValue: " + fmt.Sprint(ret.Get(0)) + "\n")
	ret.TypeCode = getTypeCode(ret.Get(0))
	Log("NewArrayValue: getTypeCode said " + fmt.Sprint(ret.TypeCode) + "\n")
	if ret.TypeCode == CUSTOM {
		ret.CustomTy = vf.GetCustomStructType(reflect.TypeOf(ret.Get(0)))
	}
	return ret
}

func ToArrayValue(o interface{}, vf ValueFactory) *ArrayValue {
	av := NewArrayValue(o, vf)
	return av
}

func FromArrayValue(av *ArrayValue) interface{} {
	return  av.ArrayVal.Interface()
}

func (ar *ArrayValue) Size() int {
	return ar.ArrayVal.Len()
}

func (ar *ArrayValue) Get(i int) interface{} {
	return ar.ArrayVal.Index(i).Interface()
}

func (ar *ArrayValue) Set(i int, value interface{}) {
	ar.ArrayVal.Index(i).Set(reflect.ValueOf(ToAssignableValue(ar.ArrayVal.Index(i),value)))
}


func getTypeCode(val interface{}) int8 {
	if val == nil {
		return NULL
	}

	switch a := val.(type) {
	case []byte:
		return BYTES
	}

	if reflect.TypeOf(val).Kind() == reflect.Slice {
		return ARRAY
	}

	switch a := val.(type) {
	case int8:
		return BYTE
	case int16:
		return SHORT
	case int32:
		return INT
	case int64:
		return LONG
	case float32:
		return FLOAT
	case float64:
		return DOUBLE
	case string:
		return STRING
	default:
		return CUSTOM
	}
	return NONE

}


func checkValue(val interface{}) int8 {
	if val == nil {
		return NULL
	}
	if val == NONE {
		return NONE
	}


	switch a := val.(type) {
	case []byte:
		return BYTES
	}

	if reflect.TypeOf(val).Kind() == reflect.Slice {
		return ARRAY
	}

	switch a := val.(type) {
	case int8:
		return checkByte(val.(int8))
	case int16:
		return checkShort(val.(int16))
	case int32:
		return checkInteger(val.(int32))
	case int64:
		return checkLong(val.(int64))
	case float32:
		return FLOAT
	case float64:
		return DOUBLE
	case string:
		return STRING
	default:
		return CUSTOM
	}
	return NONE

}


func checkLong(val int64) int8 {
	if val <= 4294967295 &&
		val >= -4294967296 {
		return checkInteger(int32(val))
	}
	return LONG
}

func checkInteger(val int32) int8 {
	if val <= 65535 &&
		val >= -65536 {
		return checkShort(int16(val))
	}
	return INT
}


func checkShort(val int16) int8 {
	if val <= 127 &&
		val >= -128 {
		return checkByte(int8(val))
	}
	return SHORT
}

func checkByte(val int8) int8 {
	if val <= MAX_TINY_INT &&
		val >= MIN_TINY_INT {
		return val
	}
	return BYTE
}

func toBytes(val string) []byte {
	return []byte(val)
}

func toByte(val interface{}) int8 {
	switch val.(type) {
	case int:
		return int8(val.(int))
	case int8:
		return val.(int8)
	case int16:
		return int8(val.(int16))
	case int32:
		return int8(val.(int32))
	case int64:
		return int8(val.(int64))
	}
	return 0
}

func toShort(val interface{}) int16 {
	switch val.(type) {
	case int:
		return int16(val.(int))
	case int8:
		return int16(val.(int8))
	case int16:
		return int16(val.(int16))
	case int32:
		return int16(val.(int32))
	case int64:
		return int16(val.(int64))
	}
	return 0
}

func toInteger(val interface{}) int32 {
	switch val.(type) {
	case int:
		return int32(val.(int))
	case int8:
		return int32(val.(int8))
	case int16:
		return int32(val.(int16))
	case int32:
		return int32(val.(int32))
	case int64:
		return int32(val.(int64))
	}
	return 0
}
